### PR TITLE
Overhaul Master Server

### PR DESCRIPTION
**Pull request by lolPants**
I don't expect this to be merged given what you told me before about not liking TypeScript, but I'm going to shoot my shot one last time. Regardless of whether this gets merged I'll still be tracking upstream changes on my fork and keeping it up-to-date, so anyone who wishes to take advantage of the new features can do so.

Improvements:
* Type safety thanks to TypeScript (I caught and fixed quite a few bugs thanks to TS)
* More structured project layout
* Optional high availability using Redis to sync multiple instances
* Multiple storage backends (SQLite for dev/single instances, or PostgreSQL for high availability)
* Optional route caching for server list using Redis
* Basic global ban system, currently you have to manually edit the DB (supersedes #8)
* Enabled CORS for server list (Supersedes #10)
* Docker Images built by CI
* Error messages for `success: false` responses

This PR also contains some opinionated changes, such as moving to Yarn v3 and enforcing a style guide.

### CODE REVIEW

**Comment by barnabwhy** [comment]
README.md is changed in a way that isn't what it should say on the actual repo.

**Comment by lolPants** [comment]
@barnabwhy If Bob decides to merge then it can be changed

File: README.md [hunk 1]
```diff
@@ -1,2 +1,11 @@
...
+> An alternate Master Server for [Northstar](https://github.com/R2Northstar/Northstar)
+
+## Why?
+This fork is built upon the [official Master Server](https://github.com/R2Northstar/NorthstarMasterServer), but maintains a number of important improvements.
+* Type safety thanks to TypeScript
+* More structured project layout
+* High availability using Redis to sync multiple instances
+* Multiple storage backends (SQLite for dev/single instances, or PostgreSQL for high availability)
+* Route Caching for commonly requested routes using Redis
+* Automatic Docker Images
```
Comment (barnabwhy) [inline README.md:11]: Change this

```suggestion
# Northstar MasterServer
Master server for [Northstar](https://github.com/R2Northstar/Northstar)

* Type safety thanks to TypeScript
* High availability using Redis to sync multiple instances
* Multiple storage backends (SQLite for dev/single instances, or PostgreSQL for high availability)
* Route Caching for commonly requested routes using Redis
* Automatic Docker Images
```

**Comment by lolPants** [comment]
I've now finished compatibility testing everything, and all works as expected. Also rebased on latest `main` to eliminate merge conflicts.

**Comment by lolPants** [comment]
This PR also comes with minor DB schema changes, you can upgrade to the new schema with the following SQL. **Please make a backup beforehand anyway.**
```sql
-- Add new column for global bans
ALTER TABLE `accounts` ADD COLUMN "isBanned" boolean NOT NULL;

-- Modify data type to BigInt to avoid overflow issues
ALTER TABLE `accounts` ADD COLUMN "authTokenExpireTime" bigint NOT NULL;
ALTER TABLE `accounts` DROP COLUMN "currentAuthTokenExpirationTime";
```

As for promoting an SQLite DB to PostgreSQL, I would recommend looking into [PGLoader](https://pgloader.readthedocs.io/en/latest/ref/sqlite.html)

**Comment by MerijnHendriks** [comment]
> what you told me before about not liking TypeScript

An understandable one. Iin my opinion typescript's "best practices" have the ability make thing unnecessarily complicated.

> Type safety thanks to TypeScript (I caught and fixed quite a few bugs thanks to TS)

If any of these bugs exist in the current codebase, could you make a separate PR for this? I think that would be beneficial to have regardless if this PR gets merged.

> - More structured project layout
> - Optional high availability using Redis to sync multiple instances
> - Multiple storage backends (SQLite for dev/single instances, or PostgreSQL for high availability)
> - Optional route caching for server list using Redis
> - Docker Images built by CI

Each of these should've been split into separate PR's.

> This PR also contains some opinionated changes, such as moving to Yarn v3 and enforcing a style guide.

Instead of enforcing someone else's styleguide, why not enforce the style @BobTheBob9 has been using in the current code?

---

My question to you is: if you maintain a project and someone comes in with a massive PR that basically rewrites your project,  would you merge it? The changes you make here might be fantasic, very beneficial or even the best way to move forward, but it breaks stability (this PR would need to be tested tremendously) and forces the existing developers to adapt to everything you introduced. In my opinion your current PR is just way too big, too opinionated, too many dependencies (increased security risk!), too many changes.

Maybe instead of rewriting everything as typescript and including such radical changes, it might be worthwhile to instead introduce these changes incrementally and start with type checking for the current javascript code through tsc and set up linting to enforce consistency with the existing style used in the current codebase.

---

Sorry if I worded things too strongly, english is not my native language.

**Comment by lolPants** [comment]
> An understandable one. Iin my opinion typescript's "best practices" have the ability make thing unnecessarily complicated.

While I agree that TypeScript can sometimes make things more verbose, in the end it forces you to write code to check the safety of your data (which is especially important with an API accepting user input), and ultimately makes the software safer and more stable as a result.

> If any of these bugs exist in the current codebase, could you make a separate PR for this? I think that would be beneficial to have regardless if this PR gets merged.

Most of these were simple typos, it would be quite a lot of work to scour through the 100+ commits on this PR to extract them all.

> Each of these should've been split into separate PR's.

While normally I would agree, a lot of the work done on my fork was a foundation that allowed me to incrementally implement new features. Moving to TypeScript allowed me to find and fix many bugs, which allowed me to safely restructure things, which allowed me to rewrite the database system, which allowed me to implement multiple data stores, etc.

> Instead of enforcing someone else's styleguide, why not enforce the style [... from] the current code?

The opinionated changes are because I am maintaining this as a fork, I intend to use my codebase to host an alternate master server. The current code has several style inconsistencies and ultimately made the code a little hard to visually parse at times. Enforcing a consistent style guide is important for the longevity of a project, and since it didn't have one before I put my own in. Happy for style opinions to be tweaked further down the line, but I would like **some** style guide to be enforced.

> My question to you is: if you maintain a project and someone comes in with a massive PR that basically rewrites your project, would you merge it?

Honestly probably not, which is why I opened this PR with the expectation that it won't be merged. However I do still believe that the changes made here are an overall benefit to the entire Northstar ecosystem, which is why I made them in the first place. Northstar as a project is really cool, but as demonstrated over the last few days the current master server implementation is beginning to crack under the pressure of so many users.

> but it breaks stability (this PR would need to be tested tremendously)

Agreed. While I have done pretty extensive testing as I've been going to ensure compatibility and feature parity, I alone can only do so much. Opening this PR was also a way to hopefully get more eyes on the code to help test.

At the end of the day, I don't mind if this PR is not merged since it is essentially my personal fork for my use case, I'd also be happy if people took the ideas I had and implemented them themselves. But for those looking to run their own master server, I intend to keep maintaining my fork, keeping it up-to-date with the official implementation and continuing to add features.

**Comment by lolPants** [comment]
This now has merge conflicts due to the new approach taken to failover, which personally I feel to be a little excessive. Sending the entirety of the server's details to the master server (for every server) every 5 seconds seems like a fast way to DDoS yourself.

Since the issue with servers being dropped after a restart of the master server is already solved by this PR by using an external storage medium for servers, I will be resolving the conflicts by eliminating the new logic. **Since the new logic on the Northstar Client should be backwards compatible with the old master server logic, nothing should break.**

**Comment by abarichello** [comment]
As much as I like most of these changes there's no way this is getting merged. Thanks for the suggestions, we'll try to test and implement some of these improvements ourselves.

**Pull request closed without merging**